src/Pos/Block/Logic.hs

```diff
@@ -222,37 +252,62 @@ classifyNewHeader (Right header) = do
 
 -- | Result of multiple headers classification.
 data ClassifyHeadersRes ssc
-    = CHsValid (BlockHeader ssc) -- ^ Header list can be applied, LCA child attached.
-    | CHsUseless !Text           -- ^ Header is useless.
-    | CHsInvalid !Text           -- ^ Header is invalid.
+    = CHsValid (BlockHeader ssc)   -- ^ Header list can be applied,
+                                   --    LCA child attached.
+    | CHsUseless !Text             -- ^ Header is useless.
+    | CHsInvalid !Text             -- ^ Header is invalid.
+    | CHsOld !EpochOrSlot !SlotId  -- ^ Newest header in the chain isn't
+                                   --    from current slot (and we're not
+                                   --    in recovery mode).
+                                   --   Fields:
+                                   --    * newest header's slot
+                                   --    * current slot
     deriving (Show)
 
 -- | Classify headers received in response to 'GetHeaders' message.
 --
 -- * If there are any errors in chain of headers, CHsInvalid is returned.
 -- * If chain of headers is a valid continuation or alternative branch,
--- lca child is returned.
+--    lca child is returned.
 -- * If chain of headers forks from our main chain too much, CHsUseless
--- is returned, because paper suggests doing so.
+--    is returned, because paper suggests doing so.
+-- * If we aren't too far behind the current slot (i.e. if 'needRecovery' is
+--    false), the newest header in the list has to be from current slot
+--    (see CSL-177), and CHsNotCurrentSlot is returned.
 classifyHeaders
     :: forall ssc m.
-       WorkMode ssc m
-    => NewestFirst NE (BlockHeader ssc) -> m (ClassifyHeadersRes ssc)
+       (SscWorkersClass ssc, WorkMode ssc m)
+    => NewestFirst NE (BlockHeader ssc)
+    -> m (ClassifyHeadersRes ssc)
 classifyHeaders headers = do
     tipHeader <- DB.getTipBlockHeader @ssc
     let tip = headerHash tipHeader
     haveOldestParent <- isJust <$> DB.getBlockHeader @ssc oldestParentHash
-    let headersValid = isVerSuccess $ verifyHeaders True (headers & _Wrapped %~ toList)
+    let headersValid = isVerSuccess $
+                       verifyHeaders True (headers & _Wrapped %~ toList)
+    needRecovery_ <- needRecovery
+    mbCurrentSlot <- getCurrentSlot
+    let newestHeaderConvertedSlot =
+            case newestHeader ^. epochOrSlotG of
+                EpochOrSlot (Left e)  -> SlotId e 0
+                EpochOrSlot (Right s) -> s
     if | not headersValid ->
-             pure $ CHsInvalid "Header chain is invalid"
+             pure $ CHsInvalid
+             "Header chain is invalid"
```

Comment:
I don't remember, probably for consistency. Or maybe there was a long string somewhere and then I deleted it and forgot to change things back. I'll revert it.